django/dkg/cancer/model_api/model.py

```diff
@@ -0,0 +1,118 @@
+import sklearn.feature_extraction
+import sklearn.preprocessing
+import scipy.sparse
+import cancer.ovr
+import cancer.models
+from sklearn.externals import joblib
+import cPickle
+
+
+def encode_features_of(articles_with_keywords_and_probas):
+    """
+    returns scipy.sparse matrix from model articles features.
+
+    :param articles_with_keywords_and_probas: pandas.DataFrame.
+    :return: sparse design matrix.
+    """
+    abstracts = sklearn.feature_extraction.text.HashingVectorizer(n_features=2**15)
+    X_abstracts = abstracts.fit_transform(articles_with_keywords_and_probas.abstract)
+
+    titles = sklearn.feature_extraction.text.HashingVectorizer(n_features=2**8)
+    X_titles = titles.fit_transform(articles_with_keywords_and_probas.title)
+
+    X = scipy.sparse.hstack((X_titles, X_abstracts))
+    return X
+
+
+def encode_labels_of(articles_with_keywords_and_probas):
+    """
+    returns scipy.sparse matrix from model article labels.
+    :param articles_with_keywords_and_probas: pandas.DataFrame
+    :return:
+    """
+    def preprocess_keywords(keywords):
+        """
+        Lower-case keyword tokens and remove dates and search sources keywords.
+
+        :param keywords: pandas.Series of (kw, proba, annotator)
+        :return: cleaned list of string keywords
+        """
+        keywords_filtered = keywords.apply(lambda kws: [kw[0].lower() for kw in kws])
+        keywords_filtered = keywords_filtered.apply(lambda kws: [kw for kw in kws if not kw.startswith('quelle,') and not kw.startswith('20')])
+        return keywords_filtered
+
+    def prune_keywords(keywords, p):
+        """
+        removes keywords that occur too infrequently.
+
+        :param keywords: pandas.Serioes of lists with article keywords.
+        :param p: cumsum relative keyword occurrence cutoff.
+        :return:
+        """
+        import itertools
+        import pandas as pd
+        import numpy as np
+
+        kws = pd.DataFrame({'kw': list(itertools.chain(*keywords))})
+        print 'kws head', kws.head()
+
+        kws_r = 1.0*kws.groupby('kw').agg({'kw': np.size}).sort_values('kw', ascending=False)/kws.shape[0]
```

Comment:
For computing normalized value histograms, this should also work. 
`kws.kw.value_counts(normalize=True)
doesn't really matter, but often pandas implements stuff efficiently.